Disambiguate between Generics in extern_spec rewriter

JonasAlaif opened the pull request:
Closes #626. The issue was caused by blindly using `syn::Generics` (generic _parameters_) as `syn::GenericArgument` (generic _arguments_).

JonasAlaif committed:
Disambiguate between Generics in extern_spec rewriter

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -226,18 +228,51 @@ pub fn generate_new_struct(item: &syn::ItemImpl) -> syn::Result<syn::ItemStruct>
+        let field: ParsableUnnamedField = match param {
+            syn::GenericParam::Type(tp) => {
+                let ident = tp.ident.clone();
+                syn::parse_quote!{ core::marker::PhantomData<#ident> }
```
Minor thing to prevent user-defined `core` imports or something…
```suggestion
                syn::parse_quote!{ ::core::marker::PhantomData<#ident> }
```
Also in L243.

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -205,6 +205,8 @@ fn rewrite_method_inputs(item_ty: &syn::Type, method: &mut ImplItemMethod) ->
     args
 }
 
+type CommaPunct<T> = syn::punctuated::Punctuated<T, syn::token::Comma>;
```
```suggestion
type CommaPunct<T> = syn::punctuated::Punctuated<T, syn::Token![,]>;
```

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -226,18 +228,51 @@ pub fn generate_new_struct(item: &syn::ItemImpl) -> syn::Result<syn::ItemStruct>
-        syn::parse_str(&format!("({})", fields_str)).unwrap();
-
-    new_struct.fields = syn::Fields::Unnamed(fields);
+    new_struct.fields = syn::Fields::Unnamed( syn::parse_quote!{ (#fields) } );
```
Does it work if you define `fields` as a `Vec<syn::Field>` and then:
```suggestion
    new_struct.fields = syn::Fields::Unnamed( syn::parse_quote!{ (#(#fields),*) } );
```
If so, you can get rid of the `CommaPunct` bit, and just have `let mut fields = vec![];`

JonasAlaif replied:
Hmm do you think it's more clear like you suggest? I could just import `syn::punctuated::Punctuated` and use `Punctuated<syn::Field, syn::Token![,]>` as the type of `fields` - that way we also do not need the type alias, but avoid this `parse_quote` black magic?

Aurel300 replied:
Sure, both save space a bit. I think this type of interpolation is used in other places already, so it wouldn't be too out of place :)

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -226,18 +228,51 @@ pub fn generate_new_struct(item: &syn::ItemImpl) -> syn::Result<syn::ItemStruct>
+        };
+        let res: syn::GenericArgument = syn::parse_quote!{#ts}; res
+    }).collect();
+    syn::parse_quote!{ < #args > }
```
(Same comment as above about `Vec`)

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -226,18 +228,51 @@ pub fn generate_new_struct(item: &syn::ItemImpl) -> syn::Result<syn::ItemStruct>
+            syn::GenericParam::Const(syn::ConstParam { ident, .. }) => ident.into_token_stream(),
+            syn::GenericParam::Lifetime(ld) => ld.lifetime.clone().into_token_stream(),
+        };
+        let res: syn::GenericArgument = syn::parse_quote!{#ts}; res
```
A `syn::parse_quote!{ #ts }` should be equivalent to `syn::parse(ts)` (or `parse2` depending on which type it is).
```suggestion
        syn::parse::<syn::GenericArgument>(ts).unwrap()
```

Aurel300 reviewed (changes requested):
Looks good! Small improvements might be possible, and more importantly: please add a test.

JonasAlaif committed:
Implement code review suggestions

JonasAlaif committed:
Add future test

 Currently is ignored since none of the stuff is implemented in Prusti

Aurel300 commented on prusti-specs/src/extern_spec_rewriter.rs:
```diff
@@ -228,27 +226,27 @@ pub fn generate_new_struct(item: &syn::ItemImpl) -> syn::Result<syn::ItemStruct>
     let generics = &new_struct.generics;
 
-    let mut fields = CommaPunct::<syn::Field>::new();
+    let mut fields = Vec::<syn::Field>::new();
```
```suggestion
    let mut fields = vec![];
```

Aurel300 commented on prusti-tests/tests/verify/pass/extern-spec/trait-bounds.rs:
```diff
@@ -0,0 +1,18 @@
+// ignore-test None of this Generics stuff currently works with Prusti...
```
Ok, could we then have a UI test as well? :D

JonasAlaif committed:
Update prusti-specs/src/extern_spec_rewriter.rs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

JonasAlaif committed:
Add UI test

The Struct Name Generator must be changed to avoid regex /[a-z0-9]{32}/ matching "PrustiStructFoo[a-z0-9]{32}" as "PrustiStructF[a-z0-9]{32}xy". It is changed in a way that preserves CamelCase, though "PrustiStructFoouuiD[a-z0-9]{32}" may not be completely clear?

JonasAlaif committed:
Fix UI test

JonasAlaif committed:
Use _ to separate Struct name and UUID

JonasAlaif committed:
Make UI test work

JonasAlaif committed:
Make the test work this time

Please "Squash and Merge" this PR

JonasAlaif committed:
Move UI test to parse dir
